### ISSUE TITLE
Inference: Also output prediction perplexity

### ISSUE THREAD

**Issue by AmitMY** [GerrySant/multimodalhugs#48]
In:
https://github.com/GerrySant/multimodalhugs/blob/0581a7cab9df315529796cbeb0615d5823a1dd03/multimodalhugs/tasks/translation/inference_utils.py#L187-L190

Implementing something like:
```py
outputs = model.generate(
    **generation_inputs,
    **gen_args,
    return_dict_in_generate=True,
    output_scores=True
)

# scores is a list of logits for each token
scores = outputs.scores  # List of tensors: [batch_size, vocab_size]
tokens = outputs.sequences[:, generation_inputs["input_ids"].shape[1]:]

log_probs = [
    F.log_softmax(score, dim=-1).gather(1, token.unsqueeze(-1)).squeeze(-1)
    for score, token in zip(scores, tokens.T)
]

log_probs_tensor = torch.stack(log_probs, dim=1)  # [batch, seq_len]
perplexity = torch.exp(-log_probs_tensor.mean())
```
would also be able to return the perplexity for each prediction. That helps with data filtering

**Comment by GerrySant** [comment]
### ✅ Resolution: Return per-sample perplexity from `inference_utils.py`

**What’s added**
- `batched_prediction(...)` now supports **per-sample perplexity** for any generation strategy (greedy / sampling / beam).
- `batched_inference(...)` returns:
  ```json
  {
    "preds": [...],
    "labels": [...], 
    "perplexities": [...]
  }
  ```
  where each `perplexities[i]` corresponds to `preds[i]`.

---

**How it works**
1. Force generation to return the information we need:
   ```python
   outputs = model.generate(
       **generation_inputs,
       **gen_args,
       return_dict_in_generate=True,
       output_scores=True,           # logits for each generated step
       pad_token_id=tokenizer.pad_token_id
   )
   ```
2. Let `T = len(outputs.scores)` be the number of generated steps, and take the **last `T` tokens** of `outputs.sequences`:
   ```python
   sequences = outputs.sequences                  # [B, L_total]
   gen_tokens = sequences[:, -T:]                 # [B, T]
   ```
3. **Teacher-forcing forward pass** to align logits with those `T` positions:
   - **Encoder–decoder**:
     ```python
     dec_inp = sequences[:, -T-1:-1]             # previous tokens (length T)
     logits = model(**enc_kwargs, decoder_input_ids=dec_inp).logits  # [B, T, V]
     ```
   - **Decoder-only**:
     ```python
     logits = model(input_ids=sequences[:, :-1]).logits[:, -T:, :]   # [B, T, V]
     ```
4. Gather chosen token log-probs and compute perplexity per sample:
   ```python
   log_probs = F.log_softmax(logits, dim=-1).gather(-1, gen_tokens.unsqueeze(-1)).squeeze(-1)  # [B, T]
   mask = (gen_tokens != tokenizer.pad_token_id).float()
   mean_logp = (log_probs * mask).sum(dim=1) / mask.sum(dim=1).clamp_min(1.0)
   perplexity = torch.exp(-mean_logp).tolist()  # one value per sample
   ```

---

**Key details & edge cases**
- Handles batches with **uniform prompts**, **no prompts**, and **variable-length prompts** (per-sample generate + pad to max length).
- If a model returns no `scores` (e.g., non-standard configs), `perplexities` falls back to `[]`.
- PAD tokens are **masked** in the average; EOS is included (common practice). If you prefer to exclude EOS as well, mask it similarly.

---

**Example return**
```python
{
  "preds": ["hello world", "bon dia"],
  "labels": ["hola món", "bon dia"],
  "perplexities": [3.21, 1.87]
}
```

---

**Status**
- Implemented and shipped (see commit `dc8a686`). ✔️

If you want this behind a flag, we can expose `return_perplexity` in the CLI/API.
